### **利用程式進行讀/寫檔案的處理**
#### [<參考網址>](https://medium.com/ccclub/ccclub-python-for-beginners-tutorial-bf0648108581)ccclub
#### **[ Python 初學第十二講—檔案處理 ]**  
利用程式進行讀/寫檔案的處理

### **檔案物件 file object**
### f = open(檔名, mode)
此時變數 f 被 assign 的是一個 file object

A file object allows us to use, access and manipulate all the user accessible files. One can read and write any such files.
(文件物件允許我們使用、存取和操作所有使用者可存取的文件。人們可以讀取和寫入任何此類文件。)

### • 開啟檔案 open()
`open(file, mode='模式')`  
參數 : 檔案名稱 & 模式(mode)

1. 檔名所要填入的是檔案路徑，指的是相對路徑，也就是相對於你的程式執行的位置而言，你的檔案在什麼地方。
    如果在同一個資料夾當中執行，可以直接寫檔案的名稱，如 "test.txt" 。

2. mode 代表要開啟檔案的模式，Python 提供了數種開啟檔案的模式，包含僅供讀取、可以覆寫檔案等等。  
以下將對於其中三種稍作簡介:

    * “r” 唯讀模式：`mode = “r”` 開啟檔案時，只能從指定的檔案讀取資料，並不能夠對這個檔案的內容進行更動。
        與此同時，如果所指定的檔案不存在，將會產生 `FileNotFoundError` 的例外。

    * “w” 寫入模式（覆寫）：如以 `mode = “w”` 開啟檔案時，意即要來進行檔案的寫入，會在開啟的位置直接覆蓋掉原本的檔案。
        如果所指定的檔案路徑/名稱不存在，會新增一個新的檔案。
        
    * “a” 寫入模式（續寫）：如以 `mode =  “a”`開啟檔案時，是指在此模式下開啟檔案要進行寫入時，會從原本的檔案最後繼續進行寫入。

### • 關閉檔案 close()
`f.close()`  
當在程式當中開啟了檔案以後，如果要停止對於這個檔案的更動或寫入，可以將檔案關閉。
當要開啟其他檔案時，當然可以直接再開啟然後取代掉目前的 file object，但是記得在使用時關閉檔案依然是一個好習慣。

========================
## 一、讀取檔案

 ### **• 讀取檔案-1**
 成功開啟檔案以後，下一個要做的就是將檔案當中的資料讀取進來

 #### `f.read()` 函式
 `file.read([size])`  
 括號當中的參數 size 是可選(optional)的，當設置了 size 的值，電腦就會自動**讀到指定的字節數量**，若沒有設置就會將整個檔案都讀取進來。

In [16]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io.txt')
words = f.read()
print(words)

f = open('./files/file_io.txt')
word = f.read(6)
print(word)
f.close()

I love Python!
Get the filename, directory, extension from a path string in Python now!!
I love


### **• 讀取檔案-2**
#### `file.readline()`
這個函式將讀取檔案中的整行資料，但是一次只讀取一行，包含 \n 字元，如果檔案當中包含了 N 行的資料，我們就必須呼叫 f.readline() N 次。

In [8]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io.txt')
word_line1 = f.readline()
word_line2 = f.readline()
print(f'line1 ==> {word_line1}',f'line2 ==> {word_line2}',sep = '\n')
# print(word_line1,word_line2,sep = '\n')
f.close()

line1 ==> I love Python!

line2 ==> Get the filename, directory, extension from a path string in Python now!!


### **• 讀取檔案-3**
#### `file.readlines()`
逐行讀取，提供一次讀取所有資料的方法  

`f.readlines()` 會將檔案當中的所有資料都逐行讀取進來，然後會將其回傳成為一個 `list` ，我們嘗試執行以下的程式碼，將檔案使用 `f.readlines()` 讀取進來以後，用變數 `k` 接住，然後將其印出以便查看：

In [17]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io.txt')
word_list = f.readlines()
print(word_list)
f.close()

['I love Python!\n', 'Get the filename, directory, extension from a path string in Python now!!']


#### **file.readlines()**
`f.readlines()` 會將檔案當中每一行當作是一個字串，然後存進串列當中回傳。因此 `for loop`經常與 `f.readlines()` 搭配使用，方法如下：

執行結果如下，由於當 `f.readlines()` 將每一行的文字讀取進來時，會連換行符號的 `\n` 一起讀進來，因此，在使用 `print()` 將其印出時，會有原本字串當中的 `\n` 以及 `print()` 本身預設的 end 字元 \n ，所以中間會出現空行：

In [18]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io.txt')

for line in f.readlines():
    print(line)
f.close()

I love Python!

Get the filename, directory, extension from a path string in Python now!!


除此之外， Python 可在不使用 `f.readlines()` 的情況下，我們可以直接利用 `for loop` 對於我們的 file object f 來 迭代(iterate)，方法如下：

In [19]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io.txt')
print(f)
for line in f:
    print(line)
f.close()

<_io.TextIOWrapper name='./files/file_io.txt' mode='r' encoding='cp950'>
I love Python!

Get the filename, directory, extension from a path string in Python now!!


========================
## **二、寫入檔案**
將資料寫入檔案。與讀取相同，Python 同樣也提供了幾種不同的輸出方式：

### **• 寫入檔案 -1**
#### `file.write()`
其概念相對於 `f.read()` ，參數的資料型態是字串 string

In [21]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io_1.txt','w')
f.write('Try to use file_io_1.txt\nWrite some contents\nLine1\nline2')

f = open('./files/file_io_1.txt','r')
contents = f.readlines()
for content in contents:
    print(content)

f.close()


Try to use file_io_1.txt

Write some contents

Line1

line2


### **• 寫入檔案 -2**
#### `file.writelines()`
如不想要每輸出一段東西就使用一次 `f.write()` ，Python 也提供了另一個概念和 `f.readlines()` 相對的函式：
`f.writelines()` 用法如下：
##### `file.writelines(seq)`
參數 seq 指的是 `f.writelines()` 的參數必須是一個**序列**，也就是 list 或是 tuple 這類的資料型態。  
而 `f.writelines()` 就會將序列 (sequence) 中的內容印出，舉例來說，假如要輸出和上面相同的檔案，可以的寫法如下：

In [22]:
# import os
# os.chdir('c:/youtube')
f = open('./files/file_io_2.txt','w')
sequence= ['Try to use file_io_2.txt \n','Write some contents \n','Line1 \n','line2']
f.writelines(sequence)

f = open('./files/file_io_2.txt','r')
contents = f.readlines()
for content in contents:
    print(content)

f.close()

Try to use file_io_2.txt 

Write some contents 

Line1 

line2


由此可見，雖然 `f.writelines()` 會將所給的參數 sequence 當中的內容一個一個印出，卻不會像 `print()` 一樣預設在輸出後面附上換行 `\n` ，因此，**必須在要換行的字串後，再自行加上 `\n`** 。

========================
## **三、print()**
Python 也提供了我們利用熟悉的 `print()`，就可以完成輸出到檔案的方法。
當我們今天想要將任何東西輸出時，都會直接使用 `print()` ，然後頂多更改參數 sep 或是 end 。
然而，除了這兩個參數以外，還有第三個參數 file 。
我們會得到了一個裝著和方才直接使用 print() 時一樣內容的 txt 檔案。

舉例來說，今天有一個 list 叫做 web_list ，內容如下：
`['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']`  
當我們對這個 list 直接使用 print 時，會得到如下的結果：

In [25]:
# 當我們對這個 list 直接使用 print 時，會得到如下的結果：
web_list = ['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']
print(web_list)

['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']


然而，當設定了 `file` 這個參數以後，將程式碼改成下方的樣子，就不一樣了：  
**可得到了一個裝著和方才直接使用 `print()` 時一樣內容的 txt 檔案。**

In [23]:
#然而，當我們設定了 file 這個參數以後，將程式碼改成下方的樣子，就不一樣了：
# import os
# os.chdir('c:/youtube')
f = open('./files/web_lists.txt','w')
web_list = ['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']
print(web_list,file = f)
print('=========')

f = open('./files/web_lists.txt','r')
contents = f.readlines()
for content in contents:
    print(content)

f.close()

['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']



## **四、在檔案中移動位置**
接著，如何更進一步的進行檔案的處理。當在進行檔案的讀取時，有時可能會想要將檔案從頭到尾讀取第二遍。

直接的想法 : for loop
程式碼在第一個 for loop 跑完以後，直接下一個 for loop 再將檔案的內容再輸出一次??
但第二次的 for loop 完全沒有印出東西來。
```python
f = open('sample.txt')
for line in f:
    print(line)
for line in f:
    print(line)
f.close()
```
![列印的結果](./figure/txtPrint_01.webp)

In [5]:
f = open('./files/sample.txt')
for line in f:
    print('loop_01',line)
for line in f:
    print('loop_02',line)
f.close()

loop_01 123

loop_01 456

loop_01 789


因為，當 file object 在處理檔案時，第一個 `for loop` 結束以後，這個指針隨著讀取已經到了文件的最後面。  
因此，接下來想要再使用 for loop 來讀取的話，如果我們沒有將指針移到檔案最前面，就無法得到任何東西。  
因此，以下要介紹用來移動這個指針或是獲取現在指針位置的方法：`file.seek()`

### **• 在檔案中移動位置 -1**
#### `file.seek()`
`f.seek()` ，顧名思義，就是在檔案當中尋找。  
不過這裡說的尋找指的並不是尋找某個字或是某個資料，而是尋找某個位置。
使用上面的例子，此時，若我們想要將檔案的內容完整輸出兩次，  
只需要在兩個 `for loop` 之間加上一行 `f.seek(0)` 即可：
```python
f = open('sample.txt')
for line in f:
    print(line)
f.seek(0)
for line in f:
    print(line)
f.close()
```

![列印的結果](./figure/txtPrint_02.webp)

In [6]:
f = open('./files/sample.txt')
for line in f:
    print('loop_01',line)
f.seek(0)
for line in f:
    print('loop_02',line)
f.close()

loop_01 123

loop_01 456

loop_01 789
loop_02 123

loop_02 456

loop_02 789


也就是說，當我們想要移動這個指針的位置，只要使用 `f.seek()` 就可以將指針移到目標位置。所以當我們將 `f.seek()` 當中的參數設為 `1` 時，將會移到文件當中的位置一。所以當我們在讀取時，會從第二個字元開始讀取。 
```python
f = open('sample.txt')
f.seek(0)
for line in f:
    print(line)
f.close()
```

就會輸出下面的結果：
![列印的結果](./figure/txtPrint_03.webp) 

In [7]:
f = open('./files/sample.txt')
f.seek(1)
for line in f:
    print(line)
f.close()

23

456

789


---
++++++++++++++++++++++++++++++++++++++++++++++
#### **<活用>**

第二次的 for loop 有沒有印出?

In [32]:
# import os
# os.chdir('c:/youtube')
f = open('./files/web_lists.txt','w')
web_list = ['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']
print(web_list,file = f)

f = open('./files/web_lists.txt','r')
# contents = f.readlines()
print('==== 1st ====')
for content in f:
    print(content)

print('==== 2nd =====') 
for content in f:
    print(content)

f.close()

==== 1st ====
['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']

==== 2nd =====


很顯然的，第二次的 for loop 完全沒有印出東西來。

原因是這樣子的，當在讀取或是寫入檔案時，可以想像成電腦裡面有一個指標，類似於在使用 word 或是任何的文字編輯器時的指標，依照他的位置，知道接下來，如果要打字會出現在什麼地方。

同樣的，當 file object 在處理檔案時也是一樣的概念。在上面的範例當中，第一個 for loop 結束以後，這個指針隨著讀取已經到了文件的最後面。因此，接下來想要再使用 for loop 來讀取的話，如果沒有將指針移到檔案最前面，就無法得到任何東西。
因此，以下要介紹用來移動這個指針或是獲取現在指針位置的方法。

### 在檔案中移動位置 -1
### file.seek()
第一個是 `f.seek()` ，顧名思義，就是在檔案當中尋找。不過這裡說的尋找指的並不是尋找某個字或是某個資料，而是尋找某個位置，
用法如下 :
`file.seek()`
繼續使用上面的例子，此時，若我們想要將檔案的內容完整輸出兩次，只需要在兩個 `for loop` 之間加上一行 `f.seek(0)` 即可：

In [33]:
# 也就是說，當想要移動這個指針的位置，只要使用 f.seek() 就可以將指針移到目標位置。

# import os
# os.chdir('c:/youtube')
f = open('./files/web_lists.txt','w')
web_list = ['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']
print(web_list,file = f)

f = open('./files/web_lists.txt','r')
# contents = f.readlines()
print('==== 1st ====')
for content in f:
    print(content)

print('==== 2nd =====') 
f.seek(0)     # 移動這個指針的位置到起始位置
for content in f:
    print(content)

f.close()

==== 1st ====
['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']

==== 2nd =====
['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']



或者是:儲存在一個位置後，再重複呼叫

In [34]:
# import os
# os.chdir('c:/youtube')
f = open('./files/web_lists.txt','w')
web_list = ['https://www.web_01.com/home','https://www.web_02.com/home','https://www.web_03.com/home','https://www.web_04.com/home','https://www.web_05.com/home','https://www.web_06.com/home']
print(web_list,file = f)

f = open('./files/web_lists.txt','r')
contents = f.readlines()
print('==== 1st ====')
for content in contents:
    print(content)

print('==== 2nd =====') 
for content in contents:
    print(content)

f.close()

==== 1st ====
['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']

==== 2nd =====
['https://www.web_01.com/home', 'https://www.web_02.com/home', 'https://www.web_03.com/home', 'https://www.web_04.com/home', 'https://www.web_05.com/home', 'https://www.web_06.com/home']



**活用End**  
***
==========================================================

### 在檔案中移動位置 -2
### `file.tell()`
`f.tell()` 和 `f.seek()` 是接近於相對的概念，  
所謂的 `f.tell()` ，顧名思義就是會回傳給你現在指標所在的位置。  
在讀取資料時使用的是 `f.read()` 時，`f.tell()`可以很輕鬆地知道目前指針的位置在哪裡。
然而，若要搭配 f.seek() 或是其他方法使用時，就需要`f.tell()`來知道目前確切的指針位置，與此同時也可以知道目前進行到檔案的什麼地方。
```python
f = open('sample.txt')
print(f.read(1))
print(f.readline())
print(f.tell())
f.close()
```


就會輸出下面的結果：
![列印的結果](./figure/txtPrint_04.webp) 

由此可見，使用 `f.tell()` 可以得到目前指針所在的位置，是下一行的開頭也就是 `123\n` 之後的位置 `4`。

In [8]:
# 舉例來說，先使用了 f.read() ，又使用了 f.readline() 後，若想要知道接下來會從哪個位置開始。
# 試寫程式碼如下：

import os
# os.chdir('c:\youtube')
# os.chdir('C:\Users\Lu\Documents\GitHu\__2024_01_08_python_openAPI__\00_selfStudy/files')
print(os.getcwd()) 
f = open('./files/file_io_2.txt','r')
print(f.read(1))
print(f.readline())
print(f.tell())
print(f.read())
f.close()

c:\youtube


FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/Lu/Documents/GitHub/__2024_01_08_python_openAPI__/00_selfStudy/files/file_io_2.txtfile_io_1.txt'


---
---
+++++++++++++++++++++++++++++++

## Python File(檔案) 方法
<ref> https://www.runoob.com/python/file-methods.html

### open() 方法
Python open() 方法用於開啟一個文件，並傳回文件對象，在對文件進行處理過程都需要使用到這個函數，如果該文件無法被打開，會拋出OSError。

注意：使用open() 方法一定要保證關閉檔案對象，也就是呼叫close() 方法。

open() 函數常用形式是接收兩個參數：檔案名稱(file)和模式(mode)。
### open ( file , mode = 'r' )

完整的語法格式為：
### open ( file , mode = 'r' , buffering =- 1 , encoding = None , errors = None , newline = None , closefd = True , opener = None )
==> 參數說明 :
file: 必需，檔案路徑（相對或絕對路徑）。
mode: 可選，檔案開啟模式
buffering: 設定緩衝
encoding: 一般使用utf8
errors: 報錯級別
newline: 區分換行符
closefd: 傳入的file參數類型
opener: 設定自訂開啟器，開啟器的回傳值必須是一個開啟的檔案描述子。

### mode 參數有：

模式 | 描述
t	文字模式(預設)。
x	寫入模式，新建一個文件，如果該文件已存在則會報錯。
b	二進制模式。
+	開啟一個檔案進行更新(可讀可寫)。
U	通用換行模式（不建議）。
r	以唯讀方式開啟檔案。文件的指針將會放在文件的開頭。這是預設模式。
rb	以二進位格式開啟一個檔案用於唯讀。文件指針放在文件的開頭。這是預設模式。一般用於非文字檔案如圖片等。
r+	開啟一個檔案用於讀寫。文件指針將會放在文件的開頭。
rb+	以二進位格式開啟一個檔案用於讀寫。文件指針將會放在文件的開頭。一般用於非文字檔案如圖片等。
w	開啟一個檔案只用於寫入。如該文件已存在則開啟文件，並從開頭開始編輯，即原有內容會被刪除。如該文件不存在，請建立新文件。
wb	以二進位格式開啟一個檔案只用於寫入。如果該文件已存在則開啟文件，並從開頭開始編輯，即原有內容會被刪除。如果該文件不存在，請建立新文件。一般用於非文字檔案如圖片等。
w+	開啟一個檔案用於讀寫。如果該文件已存在則開啟文件，並從開頭開始編輯，即原有內容會被刪除。如果該文件不存在，請建立新文件。
wb+	以二進位格式開啟一個檔案用於讀寫。如果該文件已存在則開啟文件，並從開頭開始編輯，即原有內容會被刪除。如果該文件不存在，請建立新文件。一般用於非文字檔案如圖片等。
a	開啟一個檔案用於追加。如果該檔案已存在，則檔案指標將會放在檔案的結尾。也就是說，新的內容將會被寫入到已有內容之後。如果該文件不存在，請建立新文件進行寫入。
ab	以二進位格式開啟一個檔案以進行追加。如果該檔案已存在，則檔案指標將會放在檔案的結尾。也就是說，新的內容將會被寫入到已有內容之後。如果該文件不存在，請建立新文件進行寫入。
a+	開啟一個檔案用於讀寫。如果該檔案已存在，則檔案指標將會放在檔案的結尾。文件開啟時會是追加模式。如果該檔案不存在，建立新檔案用於讀寫。
ab+	以二進位格式開啟一個檔案以進行追加。如果該檔案已存在，則檔案指標將會放在檔案的結尾。如果該檔案不存在，建立新檔案用於讀寫。

In [7]:
# mode = 'x' 寫入模式，新建一個文件，如果該文件已存在則會報錯。

import os
os.chdir('c:\youtube')
try:
    open('file_io_3.txt','x')
except:
    print('file existed')

file existed


### file 物件
file 物件使用open 函數來創建，下表列出了file 物件常用的函數：

序號	方法及描述
file.close()
關閉文件。關閉後文件不能再進行讀寫操作。

file.flush()
刷新檔案內部緩衝，直接把內部緩衝區的資料立刻寫入檔案, 而不是被動的等待輸出緩衝區寫入。

file.fileno()
傳回一個整數的檔案描述子(file descriptor FD 整數), 可以用在如os模組的read方法等一些底層操作上。

file.isatty()
如果檔案連接到一個終端設備回傳True，否則回傳False。

file.next()
返回文件下一行。

file.read([size])
從檔案讀取指定的位元組數，如果未給定或為負則讀取所有。

file.readline([size])
讀取整行，包括"\n" 字元。

file.readlines([sizeint])
讀取所有行並返回列表，若給定sizeint>0，則是設定一次讀取多少字節，這是為了減輕讀取壓力。

file.seek(offset[, whence])
設定文件目前位置

file.tell()
返回文件目前位置。

file.truncate([size])
截取文件，截取的位元組透過size指定，預設為目前文件位置。

file.write(str)
將字串寫入文件，傳回的是寫入的字元長度。

file.writelines(sequence)
向文件寫入一個序列字串列表，如果需要換行則要自行加入每行的換行符。

In [3]:
# 應用 -1
import os
os.chdir('c:/youtube')
f = open('web_list.txt')
webList = f.readlines()
print(webList)
f.close()

n = 0
for i in webList :
    n += 1
    print(f'web{n} ==> {i}')

['https://www.web_11.com/home\n', 'https://www.web_12.com/home\n', 'https://www.web_13.com/home\n', 'https://www.web_14.com/home\n', 'https://www.web_15.com/home\n', 'https://www.web_16.com/home\n']
web1 ==> https://www.web_11.com/home

web2 ==> https://www.web_12.com/home

web3 ==> https://www.web_13.com/home

web4 ==> https://www.web_14.com/home

web5 ==> https://www.web_15.com/home

web6 ==> https://www.web_16.com/home



In [5]:
# 應用 -2
import os
os.chdir('c:/youtube')
f = open('web_list.txt')
webList = f.read()
print(webList)
f.close()

n = 0
for i in webList :
    n += 1
    print(f'web{n} ==> {i}')

https://www.web_11.com/home
https://www.web_12.com/home
https://www.web_13.com/home
https://www.web_14.com/home
https://www.web_15.com/home
https://www.web_16.com/home

web1 ==> h
web2 ==> t
web3 ==> t
web4 ==> p
web5 ==> s
web6 ==> :
web7 ==> /
web8 ==> /
web9 ==> w
web10 ==> w
web11 ==> w
web12 ==> .
web13 ==> w
web14 ==> e
web15 ==> b
web16 ==> _
web17 ==> 1
web18 ==> 1
web19 ==> .
web20 ==> c
web21 ==> o
web22 ==> m
web23 ==> /
web24 ==> h
web25 ==> o
web26 ==> m
web27 ==> e
web28 ==> 

web29 ==> h
web30 ==> t
web31 ==> t
web32 ==> p
web33 ==> s
web34 ==> :
web35 ==> /
web36 ==> /
web37 ==> w
web38 ==> w
web39 ==> w
web40 ==> .
web41 ==> w
web42 ==> e
web43 ==> b
web44 ==> _
web45 ==> 1
web46 ==> 2
web47 ==> .
web48 ==> c
web49 ==> o
web50 ==> m
web51 ==> /
web52 ==> h
web53 ==> o
web54 ==> m
web55 ==> e
web56 ==> 

web57 ==> h
web58 ==> t
web59 ==> t
web60 ==> p
web61 ==> s
web62 ==> :
web63 ==> /
web64 ==> /
web65 ==> w
web66 ==> w
web67 ==> w
web68 ==> .
web69 ==> w
web70 ==> e
